projects/bootloader/scripts/flash_application_code.py

```diff
@@ -0,0 +1,153 @@
+    for chunk in chunked_application_code:
+        # Reset recieved keys in listener
+        flash.listener.datagram_messages = {}
+        flash.recv_boards = []
+        flash.flash_application_chunk(chunk)
+
+
+class FlashApplication:
+    """Flash Application Code Protocol"""
+
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, **kwargs):
+        self._check_kwargs(**kwargs)
+
+        self.channel = kwargs["channel"]
+        self.app_data = kwargs["application_code_data"]
+        self.board_ids = kwargs["board_ids"]
+        self.name = kwargs["name"]
+        self.git_version = kwargs["git_version"]
+
+        self.sender = DatagramSender(channel=self.channel, receive_own_messages=False)
+        self.listener = DatagramListener(self.listener_callback)
+        self.notifier = can.Notifier(self.sender.bus, [self.listener])
+
+        self.recv_boards = []
+        self.status_code = 0
+
+    def listener_callback(self, datagram: Datagram, board_id):
+        """Check whether the datagram has an OK status code"""
+        # Ignores non-zero codes from the client
+        if datagram._data != bytearray(0) and board_id != 0:
+            self.status_code = 1
+        elif board_id == 0:
+            # Mock controller board responses
+            # Only runs when testing - client id is 0
+            for node_id in self.board_ids:
+                if node_id == 0:
+                    continue
+                mock_msg = Datagram(
+                    datagram_type_id=0,
+                    node_ids=self.board_ids,
+                    data=bytearray(0))
+                self.sender.send(mock_msg, node_id)
+
+    def chunks(self, app_code: bytearray, chunk_size):
+        """Yield successive chunk_size byte-sized chunks from app_code."""
+        for i in range(0, len(app_code), chunk_size):
+            yield app_code[i:i + chunk_size]
+
+    def create_message(self, datagram_type_id, data):
+        """Creates a datagram to be sent with a DatagramSender"""
+
+        message = Datagram(
+            datagram_type_id=datagram_type_id,
+            node_ids=self.board_ids,
+            data=bytearray(data))
+        return message
+
+    def flash_protobuf(self):
+        """Flash Protobuf"""
+
+        # Construct and serialize protobuf
+        protobuf = pb2.FlashApplicationCode()
+        protobuf.name = self.name
+        protobuf.git_version = bytes(self.git_version)
+        protobuf.application_crc = zl.crc32(self.app_data)
+        protobuf.size = sys.getsizeof(self.app_data)
+        serialized = protobuf.SerializeToString()
+
+        # Send protobuf
+        message = self.create_message(OUTGOING_PROTOBUF_ID, serialized)
+        self.sender.send(message)
+
+        self.status_code = 0
+        timeout = time.time() + 5
+        self.recv_boards = set()
+        while set(self.board_ids) != self.recv_boards:
+            self.recv_boards = set(self.listener.datagram_messages.keys())
```

Comment:
It's valid but tricky that you're relying on the listener's `datagram_messages` field to get updated when it receives a message - you might want to put a comment to that effect somewhere.

Reply:
I changed it to use the board_id received by the datagram listener since I'd made that change earlier anyway